packages/lib-classifier/src/plugins/drawingTools/experimental/models/marks/FreehandLine/FreehandLine.js

```diff
@@ -3,278 +3,603 @@ import { FreehandLine as FreehandLineComponent } from '@plugins/drawingTools/com
 import { Mark } from '@plugins/drawingTools/models/marks'
 import { FreehandLineTool } from '@plugins/drawingTools/models/tools'
 import { FixedNumber } from '@plugins/drawingTools/types/'
+import { toJS } from 'mobx'
 
 const LINE_RESOLUTION = 0.5
 const MINIMUM_POINTS = 20
+const UNDO_ACTION_POINT_THRESHOLD = 20
 
 const SingleCoord = types.model('SingleCoord', {
   x: FixedNumber,
   y: FixedNumber
 })
 
+const ActionType = types.model('ActionType', {
+  type: types.string,
+  pointIndexDrag: types.number,
+  pointIndexClose: types.number,
+  pointsFromDrag: types.array(SingleCoord),
+  pointsFromClose: types.array(SingleCoord),
+  pointsSpliceCount: types.number,
+  isReversed: types.boolean,
+  throughBeginning: types.boolean,
+});
+
 const FreehandLineModel = types
-  .model('FreehandLineModel', {
-    pathX: types.array(FixedNumber),
-    pathY: types.array(FixedNumber)
-  })
+  .model('FreehandLineModel', {})
   .views((self) => ({
-    get coords() {
-      const [x] = self.pathX
-      const [y] = self.pathY
-      return { x, y }
+    get isValid() {
+      console.log('isValid NOT USED')
```

Comment:
`isValid` should return true if it isn’t needed. This returns undefined and the drawing canvas will automatically delete any invalid marks.